### PR TITLE
Load diffusers in native FP16/BF16 precision to reduce the memory usage

### PR DESCRIPTION
**Pull request by mvafin**
# What does this PR do?

Load diffusers model in native `float16` or `bfloat16` precision and apply 16bit patching for that case.
| Model | Before | After |
| -------- | ------- | ------- |
| FLUX-schnell | ~120Gb image | ~60Gb image |
| stable-diffusion-3.5-medium | ~38Gb | ~18Gb |

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

File: optimum/exporters/openvino/__main__.py [hunk 1]
```diff
@@ -332,6 +333,37 @@ class StoreAttr(object):
                 return model
 
             GPTQQuantizer.post_init_model = post_init_model
+    elif library_name == "diffusers" and is_openvino_version(">=", "2024.6"):
+        if os.path.exists(model_name_or_path):
```
Comment (eaidova) [inline optimum/exporters/openvino/__main__.py:337]:
```suggestion
        if Path(model_name_or_path).is_dir():
```

File: optimum/exporters/openvino/__main__.py [hunk 2]
```diff
@@ -332,6 +333,37 @@ class StoreAttr(object):
...
+        if model_part_name:
+            directory = os.path.join(path, model_part_name)
+            safetensors_files = [f for f in os.listdir(directory) if f.endswith(".safetensors")]
+            safetensors_file = None
+            if len(safetensors_files) > 0:
+                safetensors_file = safetensors_files.pop(0)
+                # filtering out fp16, fp8, etc variants to check the default variant first
+                while (len(safetensors_files) > 0
+                       and safetensors_file[:-len(".safetensors")].find(".") != -1):
+                    safetensors_file = safetensors_files.pop(0)
```
Comment (eaidova) [inline optimum/exporters/openvino/__main__.py:357]:
```suggestion
            path = Path(model_name_or_path)
        else:
            from diffusers import DiffusionPipeline
            path = DiffusionPipeline.download(model_name_or_path)
        model_part_name = None
        if (path / "transformer").is_dir():
            model_part_name = "transformer"
        elif (path / "unet").is_dir():
            model_part_name = "unet"
        dtype = None
        if model_part_name:
            directory = path, / model_part_name
            safetensors_files = [filename.name for filename in directory.glob("*.safetensors") if len(filename.suffixes) == 1 ] 
            safetensors_file = None
            if len(safetensors_files) > 0:
                safetensors_file = safetensors_files.pop(0)
```

File: optimum/exporters/openvino/__main__.py [hunk 3]
```diff
@@ -332,6 +333,37 @@ class StoreAttr(object):
...
+            if safetensors_file:
+                from safetensors import safe_open
+                safetensors_path = os.path.join(path, model_part_name, safetensors_file)
```
Comment (eaidova) [inline optimum/exporters/openvino/__main__.py:360]:
```suggestion
```

File: optimum/exporters/openvino/__main__.py [hunk 4]
```diff
@@ -15,6 +15,7 @@
 import gc
 import logging
 import operator
+import os
```
Comment (rkazants) [inline optimum/exporters/openvino/__main__.py:18]:
```suggestion
```

File: optimum/exporters/openvino/__main__.py [hunk 5]
```diff
@@ -332,6 +332,36 @@ class StoreAttr(object):
                 return model
 
             GPTQQuantizer.post_init_model = post_init_model
+    elif library_name == "diffusers" and is_openvino_version(">=", "2024.6"):
+        if Path(model_name_or_path).is_dir():
+            path = Path(model_name_or_path)
+        else:
+            from diffusers import DiffusionPipeline
+            path = DiffusionPipeline.download(model_name_or_path)
```
Comment (eaidova) [inline optimum/exporters/openvino/__main__.py:340]: in general case, it should be also extended with pipeline loading args from main_export (trust_remote_code, revision, subfolder and so on) that may affect downloading process

Comment (mvafin) [inline optimum/exporters/openvino/__main__.py:340]:
```suggestion
            path = DiffusionPipeline.download(model_name_or_path,
                                                                     revision=revision,
                                                                     cache_dir=cache_dir,
                                                                     token=token,
                                                                     local_files_only=local_files_only,
                                                                     force_download=force_download,
                                                                     trust_remote_code=trust_remote_code)
```

File: optimum/exporters/openvino/__main__.py [hunk 6]
```diff
@@ -332,6 +332,42 @@ class StoreAttr(object):
...
+            path = Path(model_name_or_path)
+        else:
+            from diffusers import DiffusionPipeline
+            path = DiffusionPipeline.download(model_name_or_path,
+                                                                     revision=revision,
+                                                                     cache_dir=cache_dir,
+                                                                     token=token,
+                                                                     local_files_only=local_files_only,
+                                                                     force_download=force_download,
+                                                                     trust_remote_code=trust_remote_code)
```
Comment (mvafin) [inline optimum/exporters/openvino/__main__.py:346]:
```suggestion
            path = DiffusionPipeline.download(model_name_or_path,
                                              revision=revision,
                                              cache_dir=cache_dir,
                                              token=token,
                                              local_files_only=local_files_only,
                                              force_download=force_download,
                                              trust_remote_code=trust_remote_code)
```

**Review by eaidova — APPROVED** [review]

**Comment by slyalin** [comment]
@mvafin, any chance to merge it soon? What is the blocker?

**Review by nikita-savelyevv — APPROVED** [review]
Cool! Are non-diffusion models already exported efficiently like this?

File: optimum/exporters/openvino/__main__.py [hunk 7]
```diff
@@ -332,6 +333,48 @@ class StoreAttr(object):
                 return model
 
             GPTQQuantizer.post_init_model = post_init_model
+    elif library_name == "diffusers" and is_safetensors_available() and is_openvino_version(">=", "2024.6"):
+        if Path(model_name_or_path).is_dir():
```
Comment (AlexKoff88) [inline optimum/exporters/openvino/__main__.py:337]: Can you please extract and encapsulate this code to a function with a meaningful name?

**Comment by mvafin** [comment]
> Cool! Are non-diffusion models already exported efficiently like this?

Only for text generation models. But it even beater there, since we do not use memory to load weights, we just copy them from one location on disk to another using mmap, for diffusers we couldn't achieve this yet.

File: optimum/exporters/openvino/__main__.py (hunk 7 above)
Comment (mvafin) [reply optimum/exporters/openvino/__main__.py]: done

**Review by AlexKoff88 — APPROVED** [review]

**Comment by AlexKoff88** [comment]
@mvafin, can you please resolve the conflicts to proceed with the merge.

**Comment by mvafin** [comment]
> @mvafin, can you please resolve the conflicts to proceed with the merge.

Done

**Comment by AlexKoff88** [comment]
@echarlaix, @IlyasMoutawwakil, the PR is ready for your review.

**Review by IlyasMoutawwakil — APPROVED** [review]

**Pull request merged**